service/history/replication/task_processor.go

```diff
@@ -175,41 +175,50 @@ func (p *taskProcessorImpl) Stop() {
 }
 
 func (p *taskProcessorImpl) processorLoop() {
-	defer func() {
-		p.logger.Debug("Closing replication task processor.", tag.ReadLevel(p.lastRetrievedMessageID))
-	}()
-
-Loop:
 	for {
-		// for each iteration, do close check first
-		select {
-		case <-p.done:
-			p.logger.Debug("ReplicationTaskProcessor shutting down.")
-			return
-		default:
+		stopping := p.processReplicationTasks()
+		if stopping {
+			break
 		}
+	}
+}
 
-		respChan := p.sendFetchMessageRequest()
-
-		select {
-		case response, ok := <-respChan:
-			if !ok {
-				p.logger.Debug("Fetch replication messages chan closed.")
-				continue Loop
-			}
+func (p *taskProcessorImpl) processReplicationTasks() (stopping bool) {
+	defer func() {
+		e := recover()
+		if e != nil {
+			p.logger.Error("A panic was encountered in replication task processing", tag.Dynamic("error", e))
+		}
+	}()
+	// for each iteration, do close check first
+	select {
+	case <-p.done:
+		p.logger.Debug("ReplicationTaskProcessor shutting down.")
+		return true
+	default:
+	}
 
-			p.logger.Debug("Got fetch replication messages response.",
-				tag.ReadLevel(response.GetLastRetrievedMessageID()),
-				tag.Bool(response.GetHasMore()),
-				tag.Counter(len(response.GetReplicationTasks())),
-			)
+	respChan := p.sendFetchMessageRequest()
 
-			p.taskProcessingStartWait()
-			p.processResponse(response)
-		case <-p.done:
-			return
+	select {
+	case response, ok := <-respChan:
+		if !ok {
+			p.logger.Debug("Fetch replication messages chan closed.")
+			return true
```

Comment:
```suggestion
			return false
```
I think?  I believe this changed since the last round